support/k8s/k8s-workload-registrar/informerhandler.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+type InformerHandlerConfig struct {
+	Log        logrus.FieldLogger
+	Controller *Controller
+	Factory    informers.SharedInformerFactory
+}
+
+type InformerHandler struct {
+	c InformerHandlerConfig
+}
+
+func NewInformerHandler(config InformerHandlerConfig) *InformerHandler {
+	return &InformerHandler{
+		c: config,
+	}
+}
+
+func (ih *InformerHandler) Run(ctx context.Context) error {
+	podInformer := ih.c.Factory.Core().V1().Pods().Informer()
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*corev1.Pod)
+			log := ih.c.Log.WithFields(logrus.Fields{
+				"ns":  pod.Namespace,
+				"pod": pod.Name,
+			})
+			if err := ih.c.Controller.SyncPod(ctx, pod); err != nil {
+				log.WithError(err).Error("Failed to sync pod")
+			}
+		},
+
+		UpdateFunc: func(old, new interface{}) {
+			oldPod := old.(*corev1.Pod)
+			newPod := new.(*corev1.Pod)
+			log := ih.c.Log.WithFields(logrus.Fields{
+				"ns":  newPod.Namespace,
+				"pod": newPod.Name,
+			})
+			if oldPod.ResourceVersion != newPod.ResourceVersion {
+				if err := ih.c.Controller.SyncPod(ctx, newPod); err != nil {
+					log.WithError(err).Error("Failed to sync pod")
+				}
+			}
+		},
+
+		DeleteFunc: func(obj interface{}) {
+			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				// This might be stale, but we're only going to use the name anyway
+				obj = tombstone.Obj
+			}
+			pod := obj.(*corev1.Pod)
+			log := ih.c.Log.WithFields(logrus.Fields{
+				"ns":  pod.Namespace,
+				"pod": pod.Name,
+			})
+			if err := ih.c.Controller.DeletePod(ctx, pod); err != nil {
+				log.WithError(err).Error("Failed to sync pod")
+			}
+		},
+	})
+
+	ih.c.Factory.Start(ctx.Done())
+
+	if ok := cache.WaitForCacheSync(ctx.Done(), podInformer.HasSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
```

Comment:
Sorry for the noob question, but exactly what caches are we waiting on here?

Reply:
This is in the guts of the kubernetes informer architecture. What it means in practice is: all currently existing pods have been loaded from the apiserver, so we've fetched a complete copy of the state of the world; before this point, the informer cache is incomplete.

I didn't do much after this, so it's not strictly needed...

Reply:
Here is a helpful diagram of what's going on behind the scenes: https://github.com/kubernetes/sample-controller/blob/master/docs/controller-client-go.md

(Annoyingly, it's hard to work with client-go without understanding this)

Reply:
I remembered why I added this, and added a log message after it. It mostly serves to identify in the log when we completed the initial sync, for debugging the cases where that sync gets stuck and never completes (indicating the apiserver is broken - and this has happened to me far too often).

Reply:
Ah ok, this is super helpful. Should we then log something prior to taking this action so if we hang here there is some indication of exactly where we hung? e.g. `ih.c.Log.Info("Syncing pods...")`. Reports or FAQ entries describing "stuck at syncing pods" would be a lot clearer

Reply:
Done